File: src/main/scala/info/bethard/timenorm/SynchronousParser.scala
package info.bethard.timenorm

import scala.collection.immutable.{ Seq, IndexedSeq }
import scala.collection.mutable.Buffer
import org.threeten.bp.temporal.ChronoUnit
import org.threeten.bp.temporal.ChronoField
import scala.collection.mutable.ListBuffer

class SynchronousParser(grammar: SynchronousGrammar) extends (Seq[String] => SynchronousParser.Tree.NonTerminal) {

  import SynchronousParser._

  def apply(sourceTokens: Seq[String]): Tree.NonTerminal = this.parse(sourceTokens)

  def parse(sourceTokens: Seq[String]): Tree.NonTerminal = {
    val chart = this.parseChart(sourceTokens)
    chart(sourceTokens.size)(0).completes match {
      case Buffer(parse) => parse.toTargetTree
      case Buffer() => {
        val nTokens = sourceTokens.size
        throw new UnsupportedOperationException("Could not parse %s. Partial parses: %s".format(
          sourceTokens.toString,
          for (size <- 1 to nTokens; start <- 0 until (nTokens - size + 1); complete <- chart(size)(start).completes) yield {
            "%s(%s)".format(complete.rule.symbol, sourceTokens.slice(start, start + size).mkString(","))
          }))
      }
      case parses => throw new UnsupportedOperationException("Found multiple parses: " + parses)
    }
  }

  def parseAll(sourceTokens: Seq[String]): Seq[Tree.NonTerminal] = {
    val chart = this.parseChart(sourceTokens)
    chart(sourceTokens.size)(0).completes.map(_.toTargetTree).toIndexedSeq
  }

  private def parseChart(sourceTokens: Seq[String]): Array[Array[ChartEntry]] = {
    val nTokens = sourceTokens.size
    val chart = Array.tabulate(nTokens + 1, nTokens) {
      (size, start) => if (size == 0 || start + size > nTokens) null else ChartEntry()
    }

    // special handling of [Number]: pass through tokens that are numbers 
    for (start <- 0 until nTokens) {
      val token = sourceTokens(start)
      if (SynchronousGrammar.isNumber(token)) {
        for (symbol <- grammar.symbolsForNumber(token.toInt)) {
          val rule = SynchronousGrammar.Rule(symbol, IndexedSeq(token), IndexedSeq(token), Map.empty)
          chart(1)(start).completes += Parse(rule, IndexedSeq.empty)
        }
      }
    }

    // fill rules that start with terminals
    for (size <- 1 to nTokens; start <- 0 until (nTokens - size + 1)) {
      val entry = chart(size)(start)
      for (rule <- grammar.sourceSeqStartsWith(sourceTokens.slice(start, start + size))) {
        if (rule.sourceSeq.size == size) {
          entry.completes += Parse(rule, IndexedSeq.empty)
        } else {
          entry.partials += PartialParse(rule, size, IndexedSeq.empty)
        }
      }
    }

    // fill in the chart from the smallest sizes to the biggest sizes
    for (size <- 1 to nTokens; start <- 0 to (nTokens - size)) {
      val entry = chart(size)(start)

      // look for ways to create entries of size `size` from the current partial parses
      for (size1 <- 1 until size) {
        val start2 = start + size1
        val size2 = size - size1
        for (partial <- chart(size1)(start).partials) {

          // partials that can be advanced to `size` using terminals
          val newSourceSeqIndex = partial.sourceSeqIndex + size2
          val symbolSeq = partial.rule.sourceSeq.slice(partial.sourceSeqIndex, newSourceSeqIndex)
          val tokenSeq = sourceTokens.slice(start2, start2 + size2)
          if (symbolSeq.forall(SynchronousGrammar.isTerminal) && symbolSeq == tokenSeq) {
            if (partial.rule.sourceSeq.size == newSourceSeqIndex) {
              entry.completes += Parse(partial.rule, partial.nonTerminalRules)
            } else {
              entry.partials += PartialParse(partial.rule, newSourceSeqIndex, partial.nonTerminalRules)
            }
          }

          // partials that can be advanced to `size` using completed non-terminals
          for (complete <- chart(size2)(start2).completes) {
            if (partial.rule.sourceSeq(partial.sourceSeqIndex) == complete.rule.symbol) {
              val sourceSeqIndex = partial.sourceSeqIndex + 1
              val nonTerminalRules = partial.nonTerminalRules :+ complete
              if (partial.rule.sourceSeq.size == sourceSeqIndex) {
                entry.completes += Parse(partial.rule, nonTerminalRules)
              } else {
                entry.partials += PartialParse(partial.rule, sourceSeqIndex, nonTerminalRules)
              }
            }
          }
        }
      }

      // create parses for rules that start with any of the currently complete parses
      // NOTE: we have to use a while-loop here because the loop itself may add more completed
      // rules that we then also need to process
      var i = 0
      while (i < entry.completes.size) {
        val complete = entry.completes(i)
        for (rule <- grammar.sourceSeqStartsWith(complete.rule.symbol)) {
          if (rule.sourceSeq.tail.isEmpty) {
            entry.completes += Parse(rule, IndexedSeq(complete))
          } else {
            entry.partials += PartialParse(rule, 1, IndexedSeq(complete))
          }
        }
        i += 1
      }
    }
    chart
  }
}

object SynchronousParser {
  
  sealed trait Tree
  object Tree {
    case class Terminal(token: String) extends Tree
    case class NonTerminal(rule: SynchronousGrammar.Rule, children: List[Tree]) extends Tree
  }

  private[SynchronousParser] case class Parse(
      rule: SynchronousGrammar.Rule,
      nonTerminalRules: IndexedSeq[Parse]) {
    
    def toTargetTree: Tree.NonTerminal = {
      var nonTerminalIndex = -1
      val children = for ((token, i) <- this.rule.targetSeq.zipWithIndex) yield {
        if (SynchronousGrammar.isTerminal(token)) {
          Tree.Terminal(token)
        } else {
          nonTerminalIndex += 1
          val nonTerminalRulesIndex = this.rule.nonTerminalAlignment(nonTerminalIndex)
          this.nonTerminalRules(nonTerminalRulesIndex).toTargetTree
        }
      }
      Tree.NonTerminal(rule, this.insertSubtreesFromParentheses(children.iterator))
    }
    
    private def insertSubtreesFromParentheses(trees: Iterator[Tree]): List[Tree] = {
      if (trees.isEmpty) {
        Nil
      } else {
        val tree = trees.next match {
          case Tree.Terminal("(") => this.parseSubtreeFollowingOpenParentheses(trees)
          case tree => tree
        }
        tree :: this.insertSubtreesFromParentheses(trees)
      }
    }
    
    private def parseSubtreeFollowingOpenParentheses(trees: Iterator[Tree]): Tree = {
      val Tree.Terminal(symbol) = trees.next
      val children = ListBuffer.empty[Tree]
      var getNext = true
      while (getNext) {
        trees.next match {
          case Tree.Terminal(")") => getNext = false
          case Tree.Terminal("(") => children += this.parseSubtreeFollowingOpenParentheses(trees)
          case tree => children += tree
        }
      }
      val rule = SynchronousGrammar.Rule("[" + symbol + "]", IndexedSeq.empty, IndexedSeq.empty, Map.empty)
      Tree.NonTerminal(rule, children.toList)
    }
  }

  private[SynchronousParser] case class PartialParse(
    rule: SynchronousGrammar.Rule,
    sourceSeqIndex: Int,
    nonTerminalRules: IndexedSeq[Parse])

  private[SynchronousParser] case class ChartEntry(
    completes: Buffer[Parse] = Buffer.empty,
    partials: Buffer[PartialParse] = Buffer.empty)
}

File: src/main/python/synchronous_parser.py
import re
from collections import deque

import synchronous_grammar as sg


class SynchronousParser(object):

    def __init__(self, grammar):
        self.grammar = grammar

    def parse_chart(self, source_tokens):
        num_tokens = len(source_tokens)
        chart = []
        for size in range(num_tokens + 1):
            chart_row = []
            for start in range(num_tokens):
                if size == 0 or start + size > num_tokens:
                    chart_row.append(None)
                else:
                    chart_row.append(_ChartEntry())
            chart.append(chart_row)

        # special handling of [Number]: pass through tokens that are numbers
        for start in range(num_tokens):
            token = source_tokens[start]
            if sg.is_number(token):
                for symbol in self.grammar.source_symbols_for_number(int(token)):
                    rule = sg.Rule(symbol, [token], [token], {})
                    chart[1][start].completes.append(_Parse(rule, []))

        # fill rules that start with terminals
        for start in range(num_tokens):
            for rule in self.grammar.source_seq_startswith(source_tokens[start]):
                initial_terminals = []
                sources_iterator = iter(rule.sources)
                next_source = next(sources_iterator)
                while sg.is_terminal(next_source):
                    try:
                        initial_terminals.append(next_source)
                        next_source = next(sources_iterator)
                    except StopIteration:
                        break
                size = len(initial_terminals)
                if source_tokens[start: start + size] == initial_terminals:
                    entry = chart[size][start]
                    if len(rule.sources) == size:
                        entry.completes.append(_Parse(rule, []))
                    else:
                        entry.partials.append(_PartialParse(rule, size, []))

        # fill in the chart from the smallest sizes to the biggest size
        for size in range(1, num_tokens + 1):
            for start in range(0, num_tokens - size + 1):
                entry = chart[size][start]

                # look for ways to create entries of size `size` from the current partial parses
                for size_1 in range(1, size):
                    start_2 = start + size_1
                    size_2 = size - size_1
                    for partial in chart[size_1][start].partials:

                        # partials that can be advanced to `size` using terminals
                        new_source_seq_index = partial.source_seq_index + size_2
                        symbols = partial.rule.sources[partial.source_seq_index: new_source_seq_index]
                        tokens = source_tokens[start_2: start_2 + size_2]
                        if all(map(sg.is_terminal, symbols)) and symbols == tokens:
                            if len(partial.rule.sources) == new_source_seq_index:
                                entry.completes.append(_Parse(partial.rule, partial.non_terminal_rules))
                            else:
                                entry.partials.append(_PartialParse(partial.rule, new_source_seq_index, partial.non_terminal_rules))

                        # partials that can be advanced to `size` using completed non-terminals
                        for complete in chart[size_2][start_2].completes:
                            if partial.rule.sources[partial.source_seq_index] == complete.rule.symbol:
                                source_seq_index = partial.source_seq_index + 1
                                non_terminal_rules = partial.non_terminal_rules + [complete]
                                if len(partial.rule.sources) == source_seq_index:
                                    entry.completes.append(_Parse(partial.rule, non_terminal_rules))
                                else:
                                    entry.partials.append(_PartialParse(partial.rule, source_seq_index, non_terminal_rules))

                    # expand complete parses if there are Nil parses beside them
                    for complete_1 in chart[size_1][start].completes:
                        for complete_2 in chart[size_2][start_2].completes:
                            if not complete_1.rule.is_nil and complete_2.rule.is_nil:
                                entry.completes.append(complete_1)
                            elif complete_1.rule.is_nil and not complete_2.rule.is_nil:
                                entry.completes.append(complete_2)

                # create parses for rules that start with any of the currently complete parses
                # NOTE: we have to use a queue here because the loop itself may add more completed
                # rules that we then also need to process
                queue = deque(entry.completes)
                while bool(queue):
                    complete = queue.popleft()
                    for rule in self.grammar.source_seq_startswith(complete.rule.symbol):
                        if not bool(rule.sources[1:]):
                            complete_2 = _Parse(rule, [complete])
                            queue.append(complete_2)
                            entry.completes.append(complete_2)
                        else:
                            entry.partials.append(_PartialParse(rule, 1, [complete]))

        return chart

    def parse_all(self, source_tokens):
        if not bool(source_tokens):
            raise TypeError("Cannot parse empty token sequence")
        chart = self.parse_chart(source_tokens)
        completes = chart[len(source_tokens)][0].completes
        roots = list(filter(lambda c: c.rule.symbol in self.grammar.root_symbols, completes))
        trees = [root.to_target_tree() for root in roots]
        if not bool(trees):
            num_tokens = len(source_tokens)
            completes = []
            for size in range(1, num_tokens + 1):
                for start in range(0, num_tokens - size):
                    for complete in chart[size][start].completes:
                        complete = "{}({})".format(complete.rule.symbol, ",".join(source_tokens[start: start + size]))
                        completes.append(complete)
            message = "Could not parse {}. Partial parses:\n{}"
            raise ValueError(message.format(source_tokens, "\n".join(completes)))
        return trees


class _Parse(object):

    def __init__(self, rule, non_terminal_rules):
        self.rule = rule
        self.non_terminal_rules = non_terminal_rules

    def to_target_tree(self):
        non_terminal_index = -1
        children = []
        for token_idx, token in enumerate(self.rule.targets):
            if sg.is_terminal(token):
                children.append(Terminal(token))
            else:
                non_terminal_index += 1
                non_terminal_rules_index = self.rule.non_terminal_alignment[non_terminal_index]
                children.append(self.non_terminal_rules[non_terminal_rules_index].to_target_tree())
        subtrees = self._insert_subtrees_from_parentheses(children)
        return NonTerminal(self.rule, subtrees)

    def _insert_subtrees_from_parentheses(self, trees):
        if not bool(trees):
            return []
        else:
            next_tree, *trees = trees
            if next_tree == Terminal("("):
                tree = self._parse_subtree_following_open_parentheses(trees)
            else:
                tree = next_tree
            return [tree] + self._insert_subtrees_from_parentheses(trees)

    def _parse_subtree_following_open_parentheses(self, trees):
        next_tree, *trees = trees
        children = []
        get_next = True
        while get_next:
            if next_tree == Terminal(")"):
                get_next = False
            elif next_tree == Terminal("("):
                children.append(self._parse_subtree_following_open_parentheses(trees))
            else:
                children.append(next_tree)
        rule = sg.Rule("[" + next_tree.symbol + "]", [], [], {})
        return NonTerminal(rule, children)


class Tree(object):
    pass


class Terminal(Tree):
    def __init__(self, token):
        self.token = token

    def __eq__(self, other):
        if isinstance(other, Terminal) and self.token == other.token:
            return True
        else:
            return False


class NonTerminal(Tree):
    def __init__(self, rule, children):
        self.token = rule
        self.children = children


class _PartialParse(object):
    def __init__(self, rule, source_seq_index, non_terminal_rules):
        self.rule = rule
        self.source_seq_index = source_seq_index
        self.non_terminal_rules = non_terminal_rules


class _ChartEntry(object):
    def __init__(self):
        self.completes = []
        self.partials = []


def to_digits(tree):
    digits = []
    if isinstance(tree, Terminal):
        # return [int(tree.token)]
        # return [tree.token]
        return tree.token
    elif isinstance(tree, NonTerminal):
        for child in tree.children:
            # digits.extend(to_digits(child))
            digits.append(to_digits(child))
    return digits


# parser = SynchronousParser(sg.load_grammar("../resources/org/clulab/timenorm/en.numbers.grammar"))
parser = SynchronousParser(sg.load_grammar("/home/egoitz/Home/Code/scala/geonorm/src/main/resources/en.grammar"))
#result = parser.parse_all(["the", "first", "week", "of", "1976"])

import sys
instr = sys.argv[1]
#instr = "the eastern slopes of the SHP , from SHP to SHP"
result = parser.parse_all(instr.split(" "))
result = to_digits(result[0])
print(result)
# result = to_digits(result[0])
# result.reverse()
# sum = 0
# for magnitude, digit in enumerate(result):
#     sum += 10 ** magnitude * digit
# print(sum)
